### ISSUE TITLE
[Question] Envoy WASM gRPC bidi streaming from a HTTP filter

### ISSUE THREAD

**Issue by NomadXD** [envoyproxy/envoy#15712]
*Title*: *Envoy WASM gRPC bidi streaming from a HTTP filter*

*Description*:
TLDR : How can we create a gRPC bidi stream **from a filter** and use the created bidi stream to send messages **when by the design of proxy-wasm-cpp-sdk won't allow us to have a pointer to the `GrpcStreamHandler` in our filter** ? 

My use case is fairly simple and straightforward. I just want to create a gRPC bidi stream when `onRequestHeaders()` is called and use the created stream to send messages in `onRequestBody()`. In order to do that I should be able to keep a pointer to the `GrpcStreamHandler` in my filter which has the methods `send()`, `onReceive()`, `onRemoteClose`. But the issue is the method which creates the gRPC bidi stream (grpcStreamHandler()), needs `std::unique_ptr<GrpcStreamHandlerBase>` as a parameter. So if I instantiate a `GrpcStreamHandler` from my filter , i have to pass it as a unique_ptr which I don't have ownership from my filter.

The following code snippet explains what I mean. 
```cpp
FilterHeadersStatus ExampleContext::onRequestHeaders(uint32_t, bool) {
  
// This is where I instantiate the handler. 
  this->stream_handler_ = std::unique_ptr<GrpcStreamHandlerBase>(new MyGrpcStreamHandler(this));
 
// And then I instantiate the grpc stream here
  auto res1 = root()->grpcStreamHandler(grpc_service_string, EchoServerServiceName, SayHelloBidiStream, initial_metadata,std::move(this->stream_handler_ ) );
  if (res1 != WasmResult::Ok) {
    LOG_ERROR("Calling gRPC server failed: " + toString(res1));
  }else{
    this->is_stream_ = true;
    LOG_INFO(std::string("gRPC stream initiated"));     
  }                 
  return FilterHeadersStatus::Continue;
}

``` 

**As a work around I'm using it as follows but it's ugly code and not safe with raw pointers** 

This is the `onRequestHeader` where I create the gRPC bidi stream. 
```cpp

FilterHeadersStatus ExampleContext::onRequestHeaders(uint32_t, bool) {
 
// This is where I instantiate the handler. 
  this->stream_handler_ =  new MyGrpcStreamHandler(this));

  auto res1 = root()->grpcStreamHandler(grpc_service_string, EchoServerServiceName, SayHelloBidiStream, initial_metadata, std::unique_ptr<GrpcStreamHandlerBase>(this->stream_handler_) );
  if (res1 != WasmResult::Ok) {
    LOG_ERROR("Calling gRPC server failed: " + toString(res1));
  }else{
    this->is_stream_ = true;
    LOG_INFO(std::string("gRPC stream initiated"));     
  }                 
  return FilterHeadersStatus::Continue;
}

```
This is `onRequestBody` where I use the created stream to send messages. 

```cpp

FilterDataStatus ExampleContext::onRequestBody(size_t body_buffer_length,
                                               bool /* end_of_stream */) {
  
  std::string sample_string = "Hello !";
  EchoRequest request;
  request.set_name(sample_string);
  if(this->is_stream_ == true){
    LOG_INFO(std::string("stream available sending message"));
// using the stream created, i try to send a message. 
    auto res = this->stream_handler_->send(request, false);
    if (res != WasmResult::Ok) {
      LOG_INFO(std::string("error sending gRPC >>>>>>>")+ toString(res));
    }
    LOG_INFO(std::string("grpc sent:"+ toString(res)));
  }
  return FilterDataStatus::Continue;
}

```

Apologies if this is not the correct repo to ask this question. Since there is no proper documentation, I opened a git issue in envoy-wasm-cpp-sdk and asked from the slack channel also. I couldn't get a answer there and that's why I opened the issue here. Anyway since my use case is fairly straight forward there should be a clean way of doing this ??? Thanks

**Comment by mattklein123** [comment]
cc @PiotrSikora

**Comment by PiotrSikora** [comment]
Duplicate of https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/issues/93. Closing this one, since it's an issue with the Proxy-Wasm C++ SDK.
